app/models/git_repository.rb

```diff
@@ -9,6 +9,11 @@ class GitRepository
     Rails.application.config.samson.cached_repos_dir
   end
 
+  def self.valid_ref_format?(reference)
+    IO.popen(['git', 'check-ref-format', "refs/head/#{reference}"]) { |_| }
```

Comment:
`heads` should be plural here. Are we not checking that it actually exists because that would force us to pull? Can't we hit the Github API in `valid_ref?` instead?